packages/embedder/embedder.ts

```diff
@@ -0,0 +1,250 @@
+
+/*
+ * TODO List
+ * - [ ] implement a clean-up function that re-queues items that haven't transitioned
+ *       to a completed state, or that failed
+ */
+
+export type DBInsert = {
+  [K in keyof DB]: Insertable<DB[K]>
+}
+
+const POLLING_PERIOD_SEC = 60 // How often do we try to grab the lock and re-index?
+const Q_MAX_LENGTH = 100 // How many EmbeddingIndex items do we batch in redis?
+const WORD_COUNT_TO_TOKEN_RATIO = 3.0 / 2 // We multiple the word count by this to estimate token count
+
+const {AI_EMBEDDER_ENABLED} = process.env
+const {SERVER_ID} = process.env
+
+tracer.init({
+  service: `embedder`,
+  appsec: process.env.DD_APPSEC_ENABLED === 'true',
+  plugins: false,
+  version: process.env.npm_package_version
+})
+tracer.use('pg')
+
+export const pg = getKysely()
+const dataLoader = getDataLoader() as DataLoaderWorker
+
+const redisClient = new RedisInstance(`embedder-${SERVER_ID}`)
+const redlock = new Redlock([redisClient], {
+  driftFactor: 0.01,
+  retryCount: 10,
+  retryDelay: 250,
+  retryJitter: 50,
+  automaticExtensionThreshold: 500
+})
+
+const refreshMetadata = async () => {
+  await refreshRetroDiscussionTopicsMeta(dataLoader)
+  // In the future, other sorts of objects to index could be added here...
+}
+const maybeQueueMetadataItems = async (modelManager: ModelManager) => {
+  const queueLength = await redisClient.zcard('embedder:queue')
+  if (queueLength >= Q_MAX_LENGTH) return
+  const itemCountToQueue = Q_MAX_LENGTH - queueLength
+  const configuredModels = [...modelManager.getEmbeddingModelsIter()].map((m) => m.getTableName())
+  const orgIds = await orgIdsWithFeatureFlag()
+
+  // For each configured embedding model, select rows from EmbeddingsMetadata
+  // that haven't been calculated nor exist in the EmbeddingsJobQueue yet
+  //
+  // Notes:
+  //   * `em.models @> ARRAY[v.model]` is an indexed query
+  //   * I don't love all overrides, I wish there was a better way
+  //     see: https://github.com/kysely-org/kysely/issues/872
+
+  const batchToQueue = await selectMetaToQueue(configuredModels, orgIds, itemCountToQueue)
+
+  if (!batchToQueue.length) {
+    console.log(`embedder: no new items to queue`)
+    return
+  }
+
+  const ejqHash: {
+    [key: string]: {
+      refUpdatedAt: Date
+    }
+  } = {}
+  const makeKey = (item: {objectType: string | null; refId: string | null}) =>
+    `${item.objectType}:${item.refId}`
+
+  const ejqValues = batchToQueue.map((item) => {
+    ejqHash[makeKey(item)] = {
+      refUpdatedAt: item.refUpdatedAt
+    }
+    return {
+      objectType: item.objectType,
+      refId: item.refId as string,
+      model: item.model,
+      state: 'queued' as DBInsert['EmbeddingsJobQueue']['state']
+    }
+  })
+
+  const ejqRows = await insertNewJobs(ejqValues)
+
+  ejqRows.forEach((item) => {
+    const {refUpdatedAt} = ejqHash[makeKey(item)]!
+    const score = new Date(refUpdatedAt).getTime()
+    redisClient.zadd('embedder:queue', score, item.id)
```

Comment:
-1 Could we reduce complexity by using PG xor redis for the job queue instead of both?

Reply:
Yes, we could. Can we do this in a separate PR?

Reply:
Let's merge this back to the trunk but not deploy it. I'd some help implementing the PG poller. I think it'll be rad.

I'll make sure to create a little "pre-prod" milestone we can track with a few issues I'm collecting

Reply:
Issue added to milestone. I think this is the one to do first.